lib/spack/spack/spec_index.py

```diff
@@ -0,0 +1,323 @@
+
+from llnl.util.lang import memoized
+
+import spack.database
+import spack.spec
+
+
+class ConcretizedSpec(object):
+    spec = None  # type: spack.spec.Spec
+
+    def __init__(self, spec):
+        # type: (spack.spec.Spec) -> None
+        if not spec.concrete:
+            raise TypeError('spec given to {0} must be fully concrete! was: {1}'
+                            .format(type(self).__name__, spec))
+        self.spec = spec
+
+    @classmethod
+    def from_abstract_spec(cls, spec):
+        # type: (spack.spec.Spec) -> ConcretizedSpec
+        return cls(spec.concretized())
+
+    def __eq__(self, other):
+        return isinstance(other, ConcretizedSpec) and self.spec == other.spec
+
+    def __ne__(self, other):
+        return not self == other
+
+    def into_hash(self):
+        # type: () -> ConcreteHash
+        return ConcreteHash(self.spec.dag_hash())
+
+    @property
+    def spec_string_name_hash_only(self):
+        # type: () -> str
+        return '{0}/{1}'.format(self.spec.name, self.into_hash().complete_hash)
+
+    def __repr__(self):
+        return '{0}(spec={1!r})'.format(type(self).__name__, self.spec)
+
+
+class IndexEntry(object):
+    concretized_spec = None  # type: ConcretizedSpec
+    record = None            # type: spack.database.InstallRecord
+
+    def __init__(self, concretized_spec, record):
+        # type: (ConcretizedSpec, spack.database.InstallRecord) -> None
+        self.concretized_spec = concretized_spec
+        self.record = record
+
+    def __repr__(self):
+        return '{0}(concretized_spec={1!r}, record={2!r})'.format(
+            type(self).__name__, self.concretized_spec, self.record)
+
+
+@add_metaclass(ABCMeta)
+class PartialHash(object):
+    FULL_HASH_STRING_LENGTH = 32  # type: ClassVar[int]
+
+    @abstractproperty
+    def hash_prefix(self):
+        # type: () -> str
+        pass
+
+    def __contains__(self, other):
+        # type: (PartialHash) -> bool
+        return other.hash_prefix.startswith(self.hash_prefix)
+
+    def __hash__(self):
+        # type: () -> int
+        return hash(self.hash_prefix)
+
+    def __eq__(self, other):
+        # type: (Any) -> bool
+        return isinstance(other, PartialHash) and self.hash_prefix == other.hash_prefix
+
+    def __ne__(self, other):
+        # type: (Any) -> bool
+        return not self == other
+
+
+class HashPrefix(PartialHash):
+    hash_prefix = None            # type: str
+
+    def __init__(self, hash_prefix):
+        # type: (str) -> None
+        assert len(hash_prefix) <= self.FULL_HASH_STRING_LENGTH, (
+            hash_prefix, self.FULL_HASH_STRING_LENGTH,
+        )
+        self.hash_prefix = hash_prefix.lower()
+
+    def __repr__(self):
+        return '{0}(hash_prefix={1!r})'.format(type(self).__name__, self.hash_prefix)
+
+
+class CompleteHash(PartialHash):
+    @abstractproperty
+    def complete_hash(self):
+        # type: () -> str
+        pass
+
+
+class ConcreteHash(CompleteHash):
```

Comment:
All the hash types here are losing me -- do we need them?  I think hashes can be much simpler than this...